frontend/public/components/dashboards-page/with-dashboard-resources.tsx

```diff
@@ -31,71 +30,70 @@ const mapStateToProps = (state: RootState) => ({
-      this.props.watchPrometheusQuery(query);
-    }
-
-    watchK8sResource: WatchK8sResource = resource => {
-      this.setState((state: WithDashboardResourcesState) => ({
-        k8sResources: [...state.k8sResources, resource],
-      }));
-    }
-
-    stopWatchK8sResource: StopWatchK8sResource = resource => {
-      this.setState((state: WithDashboardResourcesState) => ({
-        k8sResources: state.k8sResources.filter(r => r.prop !== resource.prop),
-      }));
-    }
-
-    render() {
-      return (
-        <Firehose resources={this.state.k8sResources}>
-          <WrappedComponent
-            watchURL={this.watchURL}
-            stopWatchURL={this.props.stopWatchURL}
-            watchPrometheus={this.watchPrometheus}
-            stopWatchPrometheusQuery={this.props.stopWatchPrometheusQuery}
-            urlResults={this.props[RESULTS_TYPE.URL]}
-            prometheusResults={this.props[RESULTS_TYPE.PROMETHEUS]}
-            watchK8sResource={this.watchK8sResource}
-            stopWatchK8sResource={this.stopWatchK8sResource}
-          />
-        </Firehose>
-      );
-    }
-  };
-
-export const withDashboardResources = compose(connect(mapStateToProps, mapDispatchToProps), WithDashboardResources);
+class WithDashboardResources_ extends React.Component<WithDashboardResourcesProps, WithDashboardResourcesState> {
+  private urls: Array<string> = [];
+  private queries: Array<string> = [];
+
+  constructor(props) {
+    super(props);
+    this.state = {
+      k8sResources: [],
+    };
+  }
+
+  shouldComponentUpdate(nextProps: WithDashboardResourcesProps, nextState: WithDashboardResourcesState) {
+    const urlResultChanged = this.urls.some(urlKey =>
+      this.props[RESULTS_TYPE.URL].getIn([urlKey, 'result']) !== nextProps[RESULTS_TYPE.URL].getIn([urlKey, 'result'])
+    );
+    const queryResultChanged = this.queries.some(query =>
+      this.props[RESULTS_TYPE.PROMETHEUS].getIn([query, 'result']) !== nextProps[RESULTS_TYPE.PROMETHEUS].getIn([query, 'result'])
+    );
+    const k8sResourcesChanged = this.state.k8sResources !== nextState.k8sResources;
+
+    return urlResultChanged || queryResultChanged || k8sResourcesChanged;
+  }
+
+  watchURL: WatchURL = (url, fetch) => {
+    this.urls.push(url);
+    this.props.watchURL(url, fetch);
+  }
+
+  watchPrometheus: WatchPrometheus = query => {
+    this.queries.push(query);
+    this.props.watchPrometheusQuery(query);
+  }
+
+  watchK8sResource: WatchK8sResource = resource => {
+    this.setState((state: WithDashboardResourcesState) => ({
+      k8sResources: [...state.k8sResources, resource],
+    }));
+  }
+
+  stopWatchK8sResource: StopWatchK8sResource = resource => {
+    this.setState((state: WithDashboardResourcesState) => ({
+      k8sResources: state.k8sResources.filter(r => r.prop !== resource.prop),
+    }));
+  }
+
+  render() {
+    return (
+      <Firehose resources={this.state.k8sResources}>
+        {inject(this.props.children, {
```

Comment:
It looks like we'll lose type checking with this change :(

Reply:
Yep :( I wanted to do this to keep the type checking
```suggestion
   <WithDashboardResources>
    {(props: DashboardItemProps) => <DetailsCard_ {...props} />}
  </WithDashboardResources>
```

but that does not go well with inject's `React.Children.map` as that ignores function as children.

Im not sure how else to do this to keep types in `DashboardItemProps` as required

Reply:
@spadgett  Latest change keeps types as before